consensus/hotstuff/threshold_signer.go

```diff
@@ -0,0 +1,57 @@
+package hotstuff
+
+import "github.com/onflow/flow-go/crypto"
+
+type ThresholdSigner interface {
```

Comment:
I am confused here, why we have `ThresholdSigner` and `ThresholdSigAggregator` ? `ThresholdSigAggregator` is basically part of `ThresholdSigner` interface.

Reply:
They are different.

`ThresholdSigner` is to collect random beacon sig shares, which are threshold signatures, and reconstruct the random beacon group sig. 

However, `ThresholdSigAggregator` is to collect the same random beacon sig shares, and aggregate them to be used as staking signatures.

ThresholdSigAggregator has a `Aggregator` methods whereas ThresholdSigner has a `Reconstruct` method.

Also refer to the diagram below:
![image](https://user-images.githubusercontent.com/811374/128727884-74068a42-e8ee-445a-98ec-f3af332b931e.png)

Reply:
Still not clear, sorry. I see: 
`RandomBeaconReconstructor` - reconstructs beacon sig 
`ThresholdSigAggregator` - aggregates signature from threshold signature shares
`StakingSigAggregator` - aggregates signature from staking sigs.
As I understand those are 3 interfaces that are referred in picture you have sent.
We also have `ThresholdSigner`, what is that for?

Reply:
The random beacon sig is the threshold signature share. 
It's validated once, but used twice.

The first time is to reconstruct the random beacon group signature (the reconstruction path in the above diagram from Threshold signature shares to Reconstructed group signature). This is done by the `RandomBeaconReconstructor`

The second time is to aggregate the threshold signature shares into aggregated threshold signatures. (the aggregation path in the above diagram from Threshold signature shares to Aggregate Signature). This is done by the `ThresholdSigAggregator`. 

And the path from staking signature shares to Aggregated Signature can be done in a stateless manner, and will be implemented by `StakingSigAggregator`.

The `ThresholdSigner` is the interface for [this crypto object](https://github.com/onflow/flow-go/blob/master/crypto/thresholdsign.go#L40)

Reply:
So let's say in `VoteCollector` for consensus cluster, we will use `RandomBeaconReconstructor`, `ThresholdSigAggregator`, `StakingSigAggregator` for aggregating sigs but we will use `ThresholdSigner` to verify signatures before calling `TrustedAdd`?

Reply:
The VoteCollector [only sees](https://github.com/onflow/flow-go/pull/1089/files#diff-abe94730661c5ebfd554104b47f9c67372cdca09c5ae95ad64bed79091790a2cR11-R14):
- RandomBeaconReconstructor
- ThresholdSigAggregator
- StakingSigAggregator

The VoteCollector doesn't see `ThresholdSigner`

ThresholdSigner object is the low level stateuful crypto object for verifying and storing shares, as well as reconstructing group signature.

Both SignValidator and RandomBeaconReconstructor have a reference to the ThresholdSigner object.

SignValidator uses a ThresholdSigner object to verify signature with `Verify` method
RandomBeaconReconstructor uses the same ThresholdSigner to add the verified signature with `TrustedAdd` method call.

Reply:
Ok, how `VoteCollector` will verify staking and threshold signatures? Aggregators only have functions for adding, I don't get who will do the verification.

Reply:
The `SigValidator` will [validate the signature](https://github.com/onflow/flow-go/pull/1089/files#diff-f9b5ea84da12dfb96de66c2b0852b566bce73fac11738ce0656d4017ededb266R15). 

And [`CombinedSigValidator`](https://github.com/onflow/flow-go/pull/1089/files#diff-2cb5b06735c74cc29048f9e15ea60518251fc0972b28a428a466298d0d9d4d07R21) is the implementation for consensus cluster. 

That being said, I should add a `sigValidator` field to the [ConsensusVerifyingVoteCollector object](https://github.com/onflow/flow-go/pull/1089/files#diff-abe94730661c5ebfd554104b47f9c67372cdca09c5ae95ad64bed79091790a2cR9)

Reply:
I guess this entire interface is temporary and will be replaced by the crypto library (when it's ready) right?

Reply:
The interface will be an abstraction for the crypto library, so that we can mock it in unit test.

In terms of where to place this interface, I like to place it in the package that depends on it. 

If the struct in crypto library provides methods A, B and C, but hotstuff only depends on A and B, then the ThresholdSigner interface we defined here will only have A and B. If some other module outside of hotstuff depends on B and C, they will define a new interface with B and C. If we define the interface with A B and C in the crypto library, then it would expose unnecessary method to hotstuff or other module that depends on it.

Another comment is that, given your suggestion above, we might want to rename the interface as into RandomBeaconSigner, which is implemented by the ThresholdSigner in crypto library.